smolyay/samples.py

```diff
@@ -0,0 +1,496 @@
+
+    As k does not always increase when L increases, the rate of new, unique 
+    points is capped at a rate less than or equal to :math:`2*L + 1`
+
+    The order :math:`o(L) of the slow nested Clenshaw Curtis set, which 
+    describes the number of cummulative points at each level, is used to 
+    determine the number of points at each individual level. The order of the 
+    Clenshaw Curtis set is described by the following rule:
+
+    .. math::
+
+        o(L) = \begin{cases}
+                1 & \text{ if } L = 0\\ 
+                2^{k} + 1 & \text{ if } L > 0 
+        \end{cases}
+
+        where k = \left \lceil \log_{2}(L) \right \rceil + 1
+
+    The sequence of o(L) is then :math:`{1, 3, 5, 9, 9, 17, ...}`.
+    
+    Since the points are nested, the number of unique points at
+    a level L is 
+    
+    .. math::
+        num_per_level(L) = o(L) - o(L - 1)    
+
+    Parameters
+    ----------
+    domain: list
+        Domain of the sample points.
+
+    max_level ; int
+        The maximum level the points are used for.
+    """
+
+    def _create(self):
+        r"""Create the points in the set.
+
+        Generating nested extrema of chebyshev polynomials of the first kind.
+        """
+        # create properties for levels, level 0 is a special case with 1 point
+        rule = lambda x: 1 if x == 0 else int(2 ** (numpy.ceil(numpy.log2(x)) + 1) + 1)
+        self._num_per_level = numpy.ones(self.max_level + 1, dtype=int)
+        self._num_per_level[1:] = [
+            rule(i) - rule((i - 1)) for i in range(1, self.max_level + 1)
+        ]
+        self._end_level = numpy.cumsum(self._num_per_level)
+        self._start_level = self._end_level - self._num_per_level
+
+        # points, level 0 is a special case only 0 as a point
+        points = numpy.zeros(numpy.sum(self._num_per_level))
+        degree = 0
+        for i in range(1, self.max_level + 1):
+            if self._num_per_level[i] == 0:
+                continue
+            # find indexes of extrema not already found. Fraction index/degree
+            # cannot be further simplified
+            degree = int(2**(numpy.ceil(numpy.log2(i)) + 1))
+            if i == 1:
+                # special case for level == 1
+                indexes = numpy.arange(0, degree + 1, 2, dtype=int)
+            else:
+                indexes = indexes = numpy.arange(1, degree, 2, dtype=int)
+                indexes = indexes[~(numpy.gcd(indexes, degree) > 1)]
+            points[self._start_level[i] : self._end_level[i]] = -numpy.cos(
+                numpy.pi * indexes / degree
+            )
+        self._points = self._scale_to_domain(points, [-1, 1])
+
+
+class TrigonometricPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Trigonometric sampling
+
+    The :attr:`points` for this interpolation scheme are nested
+    trigonometric points
+
+    .. math::
+
+        x^l_j = \frac{j-1}{m(l)},  1 \leq j \leq m(l), l \geq 0
```

Comment:
This documentation needs updating, it is referring to the nested version. This should probably be saying something about being the roots / extrema of a certain trig function & frequency, wherever it is these come from.

Reply:
@mzf0069 Where do these come from?

Reply:
https://arxiv.org/abs/1908.10672  

This is the reference for the trig function. I don't know what each point represents. The only note here is that the points need to be odd.

Reply:
> The only note here is that the number of points at each level needs to be odd.

Is that total number of points or unique points at each individual level? Because the total number of points is odd, the number of unique points is not.

@mphoward It's not much, but I did expand the docstrings.